rfd/0169-auto-updates-linux-agents.md

```diff
@@ -0,0 +1,364 @@
+5. Download the desired Teleport tarball specified by `agent_version` and `server_edition`.
+6. Verify the checksum.
+7. Extract the tarball to `/var/lib/teleport/versions/VERSION`.
+8. Replace any existing binaries or symlinks with symlinks to the current version.
+9. Backup /var/lib/teleport into `/var/lib/teleport/versions/OLD-VERSION/backup/teleport`
+10. Restart the agent if the systemd service is already enabled.
+11. Set `active_version` in `updates.yaml` if successful or not enabled.
+12. Replace the symlink/binary and `/var/lib/teleport` and quit (exit 1) if unsuccessful.
+13. Remove any `teleport` package if installed.
+14. Verify the symlinks to the active version still exists.
+15. Remove all stored versions of the agent except the current version and last working version.
+
+The `disable` subcommand will:
+1. Configure `updates.yaml` to set `enabled` to false.
+
+When `update` subcommand is otherwise executed, it will:
+1. Check `updates.yaml`, and quit (exit 0) if `enabled` is false, or quit (exit 1) if `enabled` is true and no proxy address is set.
+2. Query the `/v1/webapi/ping` endpoint.
+3. Check if the current time is after the time advertised in `agent_update_after`, and that `agent_auto_updates` is true.
+4. If the current version of Teleport is the latest, quit.
+5. Wait `random(0, agent_update_jitter_seconds)` seconds.
+6. Download the desired Teleport tarball specified by `agent_version` and `server_edition`.
+7. Verify the checksum.
+8. Extract the tarball to `/var/lib/teleport/versions/VERSION`.
+9. Update symlinks to point at the new version.
+10. Backup /var/lib/teleport into `/var/lib/teleport/versions/OLD-VERSION/backup/teleport`.
+11. Restart the agent if the systemd service is already enabled.
+12. Set `active_version` in `updates.yaml` if successful or not enabled.
+13. Replace the old symlink/binary and `/var/lib/teleport` and quit (exit 1) if unsuccessful.
+14. Remove all stored versions of the agent except the current version and last working version.
+
+To enable auto-updates of the updater itself, all commands will first check for an `active_version`, and reexec using the `teleport-updater` at that version if present and different.
+The `/usr/local/bin/teleport-upgrader` symlink will take precedence to avoid reexec in most scenarios.
+
+To retrieve known information about agent upgrades, the `status` subcommand will return the following:
+```json
+{
+  "agent_version_installed": "15.1.1",
+  "agent_version_desired": "15.1.2",
+  "agent_version_previous": "15.1.0",
+  "agent_edition_installed": "enterprise",
+  "agent_edition_desired": "enterprise",
+  "agent_edition_previous": "enterprise",
+  "agent_update_time_next": "2020-12-09T16:09:53+00:00",
+  "agent_update_time_last": "2020-12-10T16:00:00+00:00",
+  "agent_update_time_jitter": 600,
+  "agent_updates_enabled": true
+}
+```
+
+### Downgrades
```

Comment:
How are downgrades triggered, just `teleport-updater update SEMVER` where `SEMVER` is older than what's listed under `active_version` in `updates.yaml`? If so, might want to mention that in the move `update` section.